hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/ReplicateContainerCommandHandler.java

```diff
@@ -60,13 +61,20 @@ public ReplicateContainerCommandHandler(
   @Override
   public void handle(SCMCommand command, OzoneContainer container,
       StateContext context, SCMConnectionManager connectionManager) {
-
     final ReplicateContainerCommand replicateCommand =
         (ReplicateContainerCommand) command;
     final List<DatanodeDetails> sourceDatanodes =
         replicateCommand.getSourceDatanodes();
     final long containerID = replicateCommand.getContainerID();
 
+    DatanodeDetails dn = context.getParent().getDatanodeDetails();
+    if (dn.getPersistedOpState() !=
+        HddsProtos.NodeOperationalState.IN_SERVICE) {
+      LOG.info("Dn is of {} state. Ignore this replicate container command " +
+          "for container {}", dn.getPersistedOpState(), containerID);
+      return;
+    }
+
```

Comment:
I think this needs to go into the supervisor, so that when the task is executed by the executor, it checks to see the DN state and drops the command at that point. We could have a lot of replication commands queued which are past this point, and if the DN switches to a not IN_SERVICE state, ideally we want to prevent processing them all.

Reply:
Good point.